R/available-data-s3.R

```diff
@@ -0,0 +1,187 @@
+    if (!quiet) message("Using memory-cached available data from S3")
+    return(spod_available_data_s3_memoised(ver))
+  }
+
+  # no in-session data, check your on-disk RDS pool
+  pattern <- glue::glue("metadata_s3_v{ver}_\\d{{4}}-\\d{{2}}-\\d{{2}}\\.rds$")
+  rds_files <- fs::dir_ls(
+    path = metadata_folder,
+    type = "file",
+    regexp = pattern
+  ) |>
+    sort()
+
+  latest_file <- utils::tail(rds_files, 1)
+  latest_date <- if (length(latest_file) == 1) {
+    stringr::str_extract(basename(latest_file), "\\d{4}-\\d{2}-\\d{2}") |>
+      as.Date()
+  } else {
+    NA
+  }
+
+  needs_update <- isTRUE(force) ||
+    length(rds_files) == 0 ||
+    (!is.na(latest_date) && latest_date < Sys.Date())
+
+  if (!needs_update) {
+    if (!quiet) message("Using existing disk cache: ", latest_file)
+    return(readRDS(latest_file))
+  }
+
+  # if forcing, also clear old disk files
+  if (isTRUE(force) && length(rds_files) > 0) {
+    fs::file_delete(rds_files)
+  }
+
+  # fetch via the memoised function (this will re-hit S3 if we forgot it)
+  if (!quiet) message("Fetching latest metadata from AmazonS3 (v", ver, ")...")
+  dat <- spod_available_data_s3_memoised(ver)
+
+  # write a new RDS stamped with today's date
+  file_date <- format(Sys.Date(), "%Y-%m-%d")
+  out_path <- file.path(
+    metadata_folder,
+    glue::glue("metadata_s3_v{ver}_{file_date}.rds")
+  )
+  saveRDS(dat, out_path)
+  if (!quiet) message("Cached new data to: ", out_path)
+
+  dat
+}
+
+
+spod_available_data_s3_function <- function(
+  ver = c(1, 2)
+) {
+  ver <- as.character(ver)
+  ver <- match.arg(ver)
+
+  bucket <- paste0("mitma-movilidad-v", ver)
+
+  # original_aws_region <- Sys.getenv("AWS_DEFAULT_REGION")
+  # original_aws_url_style <- Sys.getenv("AWS_S3_URL_STYLE")
+  # on.exit({
+  #   Sys.setenv(
+  #     AWS_DEFAULT_REGION = original_aws_region,
+  #     AWS_S3_URL_STYLE = original_aws_url_style
+  #   )
+  # })
+  # Sys.setenv(
+  #   AWS_DEFAULT_REGION = "eu-west-1",
+  #   AWS_S3_URL_STYLE = "virtual"
+  # )
+
+  if (ver == 1) {
+    url_prefix <- "https://opendata-movilidad.mitma.es/"
+  } else {
+    url_prefix <- "https://movilidad-opendata.mitma.es/"
+  }
+
+  s3 <- paws.storage::s3(
+    config = list(
+      credentials = list(
+        anonymous = TRUE
+      )
+    )
+  )
+
+  all_objects <- list_objects_v2_all(s3, bucket)
+
+  # all_objects <- aws.s3::get_bucket_df(
+  #   bucket = bucket,
+  #   prefix = "", # root of bucket
+  #   max = Inf # fetch beyond the default 1000
+  # )
+
+  all_objects <- all_objects |>
+    dplyr::as_tibble() |>
+    dplyr::mutate(
+      target_url = paste0(url_prefix, .data$Key),
+      pub_ts = as.POSIXct(
+        .data$LastModified,
+        format = "%Y-%m-%dT%H:%M:%OSZ",
+        tz = "UTC"
+      ),
+      file_size_bytes = as.numeric(.data$Size),
+      etag = gsub('\\"', '', .data$ETag)
+    ) |>
+    dplyr::select(
+      .data$target_url,
```

Comment:
Why not

```suggestion
      target_url,
```

out of interest?

Reply:
This use case seems different, in our case `target_url` is the hardcoded column name, not a text string, right?

for (var in [names](https://rdrr.io/r/base/names.html)(mtcars)) {
  mtcars [%>%](https://magrittr.tidyverse.org/reference/pipe.html) [count](https://dplyr.tidyverse.org/reference/count.html)(.data[[var]]) [%>%](https://magrittr.tidyverse.org/reference/pipe.html) [print](https://rdrr.io/r/base/print.html)()
}

Reply:
@Robinlovelace 

for example, I forgot to fix the function below

```r
spod_store_etags <- function() {
  available_data <- spod_available_data(1, check_local_files = TRUE)
  available_data <- available_data |>
    dplyr::filter(downloaded == TRUE)
  local_etags <- available_data$local_path |>
    purrr::map_chr(~ spod_compute_s3_etag(.x), .progress = TRUE)
  available_data <- available_data |>
    dplyr::mutate(local_etag = local_etags) |>
    dplyr::as_tibble()
  return(available_data)
}
```

I get:

```
❯ checking R code for possible problems ... NOTE
  spod_store_etags: no visible binding for global variable ‘downloaded’
  Undefined global functions or variables:
    downloaded
```


Then if I prepend `downloaded` with bangs(is that what they called...?)/exclamation marks:

```r
dplyr::filter(!!downloaded == TRUE)
```

Same NOTE:

```
❯ checking R code for possible problems ... NOTE
  spod_store_etags: no visible binding for global variable ‘downloaded’
  Undefined global functions or variables:
    downloaded
```

Then replacing the problematic line with:
```r
dplyr::filter(.data$downloaded == TRUE)
```

No notes anymore 🤷‍♂️

Reply:
Yeah, that's a fair reason for using the `.data` syntax. 

I think messages like this 

```
  spod_store_etags: no visible binding for global variable ‘downloaded’
```

Can be resolved with `utils::globalVariables("downloaded")` somewhere in the package code, as outlined here https://forum.posit.co/t/how-to-solve-no-visible-binding-for-global-variable-note/28887/2 but in the very next post the `.data` syntax is recommended. Was just trying to understand the reasoning.